server/config/config.go

```diff
@@ -409,124 +439,146 @@ func (f *structSliceFlag) Set(value string) error {
+type Configurator struct {
+	gc          *generalConfig
+	presenceMap map[string]struct{}
 }
 
-func parseConfig(fileBytes []byte) (*generalConfig, error) {
+func NewConfiguratorFromData(data []byte) (*Configurator, error) {
 	// expand environment variables
-	expandedFileBytes := []byte(os.ExpandEnv(string(fileBytes)))
+	expandedData := []byte(os.ExpandEnv(string(data)))
 
 	// Unmarshal in strict mode once and warn about invalid fields.
 	var syntaxValidationConfig generalConfig
-	if err := yaml.UnmarshalStrict([]byte(expandedFileBytes), &syntaxValidationConfig); err != nil {
+	if err := yaml.UnmarshalStrict([]byte(expandedData), &syntaxValidationConfig); err != nil {
 		log.Warningf("Unknown fields in config: %s", err)
 	}
 
-	if err := yaml.Unmarshal([]byte(expandedFileBytes), &sharedGeneralConfig); err != nil {
+	gc := &generalConfig{}
+	if err := yaml.Unmarshal([]byte(expandedData), gc); err != nil {
+		return nil, fmt.Errorf("Error parsing config file: %s", err)
+	}
+
+	generalConfigMap := make(map[interface{}]interface{})
+	if err := yaml.Unmarshal([]byte(expandedData), &generalConfigMap); err != nil {
 		return nil, fmt.Errorf("Error parsing config file: %s", err)
 	}
 
-	// Re-parse flags so that they override the YAML config values.
-	flag.Parse()
+	// The shared config caches the last config parsed from a file so that a call to
+	// `readConfig` with an empty `configFile` can return the cached config.
+	sharedConfigurator = &Configurator{gc, populateYamlPresenceMap(generalConfigMap)}
 
-	return &sharedGeneralConfig, nil
+	return sharedConfigurator, nil
 }
 
-func readConfig(fullConfigPath string) (*generalConfig, error) {
-	if fullConfigPath == "" {
-		return &sharedGeneralConfig, nil
+func NewConfigurator(configFilePath string) (*Configurator, error) {
+	if configFilePath == "" {
+		return sharedConfigurator, nil
 	}
-	log.Infof("Reading buildbuddy config from '%s'", fullConfigPath)
+	log.Infof("Reading buildbuddy config from '%s'", configFilePath)
 
-	_, err := os.Stat(fullConfigPath)
+	_, err := os.Stat(configFilePath)
 
 	// If the file does not exist then we are SOL.
 	if os.IsNotExist(err) {
-		return nil, fmt.Errorf("Config file %s not found", fullConfigPath)
+		return nil, fmt.Errorf("Config file %s not found", configFilePath)
 	}
 
-	fileBytes, err := os.ReadFile(fullConfigPath)
+	fileBytes, err := os.ReadFile(configFilePath)
 	if err != nil {
 		return nil, fmt.Errorf("Error reading config file: %s", err)
 	}
 
-	return parseConfig(fileBytes)
+	return NewConfiguratorFromData(fileBytes)
 }
 
-type Configurator struct {
-	gc *generalConfig
-}
-
-func NewConfigurator(configFilePath string) (*Configurator, error) {
-	conf, err := readConfig(configFilePath)
-	if err != nil {
-		return nil, err
+func ParseAndReconcileFlagsAndConfig(configFilePath string) (*Configurator, error) {
+	RegisterAndParseFlags()
+	if configFilePath == "" {
+		configFilePath = *configFile
 	}
```

Comment:
I moved the use of the `config_file` flag into `ParseAndReconcileFlagsAndConfig`, so now a caller can still call `NewConfigurator` with an empty string and get a configurator using the previously parsed config, regardless of the status of the `config_file` flag. Additionally, if we wanted, we could move this function and the `config_file` flag to their own package and then only define the `config_file` flag when we wanted to use this function specifically, rather than every time we import the config.